[Feature] Decouple Diactoros from PSR7Client::acceptRequest() via PSR-17 HTTP factories

1ma opened the pull request:
With this Pull Request `PSR7Client` now accepts `ServerRequestFactoryInterface`, `StreamFactoryInterface` and `UploadedFileFactoryInterface` instances in its constructor to customize the PSR-7 implementation returned by `PSR7Client::acceptRequest()`. The goal of this change is to simplify the integration of roadrunner with several other PSR-7 frameworks, such as Slim.

All three dependencies are optional. When they are not explicitly passed `http-interop/http-factory-diactoros` is used by default.

Everything should be backwards-compatible except for one tiny detail: calling `withProtocolVersion($ctx['protocol'])` on a Diactoros `ServerRequest` breaks several http tests. I haven't figured out what's the problem yet, for now this value is left hardcoded to `1.1`.

(FTR, the [initial discussion in Reddit](https://www.reddit.com/r/PHP/comments/9k7mh2/roadrunner_php_application_server_v124_released/e70s2cz/))

wolfy-j commented:
Thank you for PR, I will carefully review it over this week as it might cause any sort of issues :) I'll try to debug 1.1 hardcoded value since we support HTTP/2 it would be sad if PHP cannot identify it.

wolfy-j commented:
I hope there are not much performance implications.

1ma commented:
Yep, there's a lot of clones now, but it turns out the PSR-17 API is very limited and forces you to rely on PSR-7 after the first instantiation... Do you have any standardized benchmarks?

wolfy-j commented:
Yep: `go test ./service/http -bench=.`

wolfy-j commented:
`BenchmarkHandler_Listen_Echo` performs basic performance test.

wolfy-j commented:
Found the bug with protocol, Golang describes the protocol using full notation `HTTP/1.1` or `HTTP/2` while Zend expects `1.0` or `2` only. But... Diactoros only has protocol validation for `with` method, not in original static constructor. I will fix it while reviewing the merge.

wolfy-j commented on src/PSR7Client.php:
```diff
@@ -53,36 +81,36 @@ public function acceptRequest()
+        if ($ctx['parsed']) {
+            $request = $request->withParsedBody(json_decode($body, true));
+        } else if ($body !== null) {
+            $bodyStream = $this->streamFactory->createStream($body);
```
This worries me, you've created stream based on body data and then wrote body into it. Either tests are not complete or json_parser ignore the second segment or something else is wrong.

1ma replied:
Lol, honest mistake (it mimicked the old code). You are right, this should have broken ~some~ a lot of tests :|

wolfy-j commented on src/PSR7Client.php:
```diff
@@ -144,8 +172,12 @@ private function wrapUploads($files): array
 
-            $result[$index] = new Diactoros\UploadedFile(
-                $f['tmpName'],
+            $stream = UPLOAD_ERR_OK === $f['error'] ?
```
Let's stick to if/else there. Ternary operator does not make it easier to understand :)

1ma replied:
Sure, no problem.

1ma commented:
WRT the benchmarks, the performance hit of these changes is not disastrous, but it is noticeable.

```
$ git checkout psr17-support 
Switched to branch 'psr17-support'
Your branch is up-to-date with 'origin/psr17-support'.

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    606305 ns/op
BenchmarkConfig_FetchIndexes-8   	  200000	      5848 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.062s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    605304 ns/op
BenchmarkConfig_FetchIndexes-8   	  200000	      5833 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.068s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    609852 ns/op
BenchmarkConfig_FetchIndexes-8   	  300000	      5494 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.539s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    605620 ns/op
BenchmarkConfig_FetchIndexes-8   	  200000	      5957 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.070s


$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    596712 ns/op
BenchmarkConfig_FetchIndexes-8   	  300000	      5770 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.555s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    597364 ns/op
BenchmarkConfig_FetchIndexes-8   	  300000	      5899 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.592s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    601763 ns/op
BenchmarkConfig_FetchIndexes-8   	  300000	      5648 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	8.036s

$ go test ./service/http -bench .
goos: linux
goarch: amd64
pkg: github.com/spiral/roadrunner/service/http
BenchmarkHandler_Listen_Echo-8   	    2000	    597632 ns/op
BenchmarkConfig_FetchIndexes-8   	  200000	      5813 ns/op
PASS
ok  	github.com/spiral/roadrunner/service/http	7.015s
```

mnavarrocarter commented:
This pr makes happy. Sorry, haven't seen this before. Feel free to close #45

wolfy-j commented:
I did not find any noticeable performance issues.

1ma commented:
Thanks! :smile:

wolfy-j commented:
Though, I have no idea why Slim does not work with PSR-7. This is bothering me a bit, the whole concept of PSR-7 is to provide compatibility between the projects...

1ma commented:
My understanding is that this issue will be solved in Slim 4.0, which has not been released yet.

https://github.com/slimphp/Slim/issues/2048
